core/src/main/java/io/questdb/griffin/engine/groupby/StableAwareStringHolder.java

```diff
@@ -76,17 +68,28 @@ public char charAt(int index) {
             // and we assume of() is called more frequently than charAt()
             long directPtr = Unsafe.getUnsafe().getLong(ptr + HEADER_SIZE);
             assert directPtr != 0;
-            assert index < length();
             return Unsafe.getUnsafe().getChar(directPtr + 2L * index);
         } else {
             return Unsafe.getUnsafe().getChar(ptr + HEADER_SIZE + 2L * index);
         }
     }
 
-    private void clear() {
-        if (ptr != 0) {
-            Unsafe.getUnsafe().putInt(ptr + LEN_OFFSET, 0);
-            direct = false;
+    public void clearAndSet(@Nullable CharSequence cs) {
+        clear();
+        if (cs instanceof StableDirectSequence) {
```

Comment:
We'll have to keep this `instanceof` since we can't tweak `CharSequence` interface (we could extend it, but that'd be ugly).